src/test/java/com/aws/greengrass/authorization/WildcardVariableTrieTest.java

```diff
@@ -0,0 +1,187 @@
+        assertFalse(rt.matches("89/1/2/3/4/90", true));
+
+        WildcardVariableTrie rt1 = new WildcardVariableTrie();
+        rt1.add("+");
+        assertTrue(rt1.matches("abc", true));
+        assertFalse(rt1.matches("/123", true));
+        rt1.add("+/+");
+        assertTrue(rt1.matches("/123", true));
+        assertFalse(rt1.matches("/123/", true));
+
+        // Test Invalid usages
+        WildcardVariableTrie rt2 = new WildcardVariableTrie();
+        rt2.add("ax/qwe+/123");
+        rt2.add("12/+23");
+        rt2.add("/+/++");
+        assertFalse(rt2.matches("ax/qwert/123", true));
+        assertFalse(rt2.matches("/33/ty", true));
+        assertFalse(rt2.matches("12/x23", true));
+        assertTrue(rt2.matches("/zz/++", true));
+
+        // Flip allowMQTT and re-test previously authorized resources
+        assertFalse(rt.matches("abc/def/123", false));
+        assertFalse(rt.matches("xyz/ghj/abcd//1", false));
+        assertFalse(rt.matches("123/56", false));
+        assertFalse(rt.matches("123/fgh/ert", false));
+        assertFalse(rt.matches("89/1/2/3/90", false));
+        assertFalse(rt.matches("89/1/2//90", false));
+        assertFalse(rt1.matches("abc", false));
+        assertFalse(rt1.matches("/123", false));
+        assertFalse(rt2.matches("/zz/++", false));
+
+    }
+
+    @Test
+    void testMultipleWildcardsMatching() {
+        WildcardVariableTrie rt = new WildcardVariableTrie();
+        rt.add("xyz*/+/*room/#");
+        rt.add("12/*#");
+        assertTrue(rt.matches("xyzzzz/89/bedroom/light/5", true));
+        assertTrue(rt.matches("12/345#", true));
+        assertFalse(rt.matches("xyzzzz/89/fan2/bedroom/light/5", true));
+        assertFalse(rt.matches("12/45/67", true));
+        rt.add("x*/+/#");
+        rt.add("a/+/+/#/+");
+        assertTrue(rt.matches("xyzzzz/89/fan2/bedroom/light/5", true));
+        assertTrue(rt.matches("xyzzz34/matchPlus", true));
+        assertTrue(rt.matches("a/b/c/#/d", true));
+        assertFalse(rt.matches("xyzzz34", true));
+        assertFalse(rt.matches("a/b/c/d/e", true));
+        // Add everything to see previous resources authorized now
+        rt.add("#");
+        assertTrue(rt.matches("xyzzzz/89/fan2/bedroom/light/5", true));
+        assertTrue(rt.matches("12/45/67", true));
+        assertTrue(rt.matches("xyzzz34", true));
+        assertTrue(rt.matches("a/b/c/d/e", true));
+
+        WildcardVariableTrie rt1 = new WildcardVariableTrie();
+        rt1.add("2131/#/#/+/ui/#");
+        assertFalse(rt1.matches("2131/x/y/z/ui/1", true));
+        assertTrue(rt1.matches("2131/#/#/z/ui/1/56", true));
+        // Add everything to see previous resources authorized now
+        rt1.add("*");
+        assertTrue(rt1.matches("2131/x/y/z/ui/1", true));
+
+        WildcardVariableTrie rt2 = new WildcardVariableTrie();
+        rt2.add("abc*/#");
+        rt2.add("x*/+/#");
+        assertTrue(rt2.matches("abc123", true));
```

Comment:
is this actually right?

Shouldn't this need to be `abc123/<something>`?

Reply:
Yes, `#` matches the previous level as well.
`abc/#` matches both `abc/<something>` and `abc`